src/NServiceBus.Transport.SQS/InputQueuePump.cs

```diff
@@ -388,6 +336,78 @@ await MovePoisonMessageToErrorQueue(receivedMessage, messageProcessingCancellati
             }
         }
 
+        public static TransportMessage ExtractTransportMessage(Message receivedMessage)
+        {
+            TransportMessage transportMessage;
+            string messageId;
+            var nativeMessageId = receivedMessage.MessageId;
+            if (receivedMessage.MessageAttributes.TryGetValue(Headers.MessageId, out var messageIdAttribute))
+            {
+                messageId = messageIdAttribute.StringValue;
+            }
+            else
+            {
+                messageId = nativeMessageId;
+            }
+
+            if (receivedMessage.MessageAttributes.TryGetValue(TransportHeaders.Headers, out var headersAttribute))
+            {
+                transportMessage = new TransportMessage
+                {
+                    Headers = JsonSerializer.Deserialize<Dictionary<string, string>>(headersAttribute.StringValue) ?? [],
+                    Body = receivedMessage.Body
+                };
+                transportMessage.Headers[Headers.MessageId] = messageId;
```

Comment:
Technically this is a bug. We can remove this line in favor of always checking for the presence of the ID in headers before setting the override ID at the end of the extraction.